pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go

```diff
@@ -0,0 +1,150 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/client-go/tools/cache"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	tenancyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
+	kcpinformers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions"
+	"github.com/kcp-dev/kcp/pkg/reconciler/apis/apibinding"
+	"github.com/kcp-dev/kcp/third_party/conditions/util/conditions"
+)
+
+const (
+	PluginName  = "apis.kcp.dev/CRDNoOverlappingGVR"
+	byWorkspace = PluginName + "-byWorkspace"
+)
+
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName,
+		func(_ io.Reader) (admission.Interface, error) {
+			return &crdNoOverlappingGVRAdmission{
+				Handler: admission.NewHandler(admission.Create),
+			}, nil
+		})
+}
+
+type crdNoOverlappingGVRAdmission struct {
+	*admission.Handler
+
+	apiBindingIndexer          cache.Indexer
+	apiBindingIndexerInitError error
+}
+
+// Ensure that the required admission interfaces are implemented.
+var _ = admission.ValidationInterface(&crdNoOverlappingGVRAdmission{})
+var _ = admission.InitializationValidator(&crdNoOverlappingGVRAdmission{})
+
+func (p *crdNoOverlappingGVRAdmission) SetKcpInformers(informers kcpinformers.SharedInformerFactory) {
+	if err := informers.Apis().V1alpha1().APIBindings().Informer().AddIndexers(cache.Indexers{byWorkspace: indexByWorkspace}); err != nil {
+		p.apiBindingIndexerInitError = err
+		return
+	}
+
+	// just in case the plugin gets init multiple times in case of an error
+	p.apiBindingIndexerInitError = nil
+	p.SetReadyFunc(informers.Apis().V1alpha1().APIBindings().Informer().HasSynced)
+	p.apiBindingIndexer = informers.Apis().V1alpha1().APIBindings().Informer().GetIndexer()
+}
+
+func (p *crdNoOverlappingGVRAdmission) ValidateInitialization() error {
+	if p.apiBindingIndexerInitError != nil {
+		return fmt.Errorf(PluginName+" plugin failed to initialize %q APIBindings indexer, err = %v", byWorkspace, p.apiBindingIndexerInitError)
+	}
+	if p.apiBindingIndexer == nil {
+		return fmt.Errorf(PluginName + " plugin needs an APIBindings indexer")
+	}
+	return nil
+}
+
+// Validate checks if the given CRD's Group and Resource don't overlap with bound CRDs
+func (p *crdNoOverlappingGVRAdmission) Validate(ctx context.Context, a admission.Attributes, _ admission.ObjectInterfaces) error {
+	if a.GetResource().GroupResource() != apiextensions.Resource("customresourcedefinitions") {
+		return nil
+	}
+	if a.GetKind().GroupKind() != apiextensions.Kind("CustomResourceDefinition") {
+		return nil
+	}
+	clusterName, err := request.ClusterNameFrom(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve cluster from context: %w", err)
+	}
+	// ignore CRDs targeting system and non-root workspaces
+	if clusterName == apibinding.ShadowWorkspaceName || clusterName == logicalcluster.New("system:admin") || !clusterName.HasPrefix(tenancyv1alpha1.RootCluster) {
```

Comment:
nit: the last test implies the other two. Am fine with either, the first two or the last :)